### PR TITLE
Orders+OrderItems

### PR DESCRIPTION
**Pull request by binh0804**
Orders + OrderItems

### CODE REVIEW

File: src/apis/controllers/orders.controller.js [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+  { customerId: 1, productId: 22, quantity: 3 },
+  { customerId: 1, productId: 23, quantity: 2 },
+  { customerId: 1, productId: 24, quantity: 1 },
+];
+
+const ordersController = {
+  //  Create new order
+  createOrder: controllerWrapper(
+    async (req, res, { errorResponse, successResponse, sql, redis }) => {
+      const { customerId, adressId, shippingFee, couponType } = req.body;
```
Comment (hualocson) [inline src/apis/controllers/orders.controller.js]: Check spell `adressId`

File: src/apis/routes/orders.routes.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import adminAuthorization from "../middlewares/auth";
+/**
+ *
+ * @param {import('express').Router} router
+ */
+
+const ordersRoutes = (router) => {
+  // create new order with customerId,status,shippingFee,total,couponId
+  router.post(
+    "/create-order",
```
Comment (hualocson) [inline src/apis/routes/orders.routes.js]: use path `/orders` method `post` for create new order

**Review by hualocson — CHANGES_REQUESTED** [review]

File: src/apis/controllers/orders.controller.js [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+            `;
+      if (!order) {
+        return errorResponse(`Order with id ${id} not found`, 404);
+      }
+      // Get all attributes of an order by id
+      const [orderById] = await sql`
+                SELECT id,customer_id,status,shipping_fee,total,coupon_id,created_at,updated_at,deleted_at,canceled_at,completed_at, delivery_at
+                FROM orders
+                WHERE id = ${id}
+            `;
```
Comment (hualocson) [inline src/apis/controllers/orders.controller.js]: Query 1 time

File: src/apis/controllers/orders.controller.js [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+      if (!customer) {
+        return errorResponse(`Customer with id ${customerId} not found`, 404);
+      }
+      // Get all attribute of orders by customerId
+      const orders = await sql`
+                SELECT id,customer_id,status,shipping_fee,total,coupon_id,created_at,updated_at,deleted_at,canceled_at,completed_at, delivery_at
+                FROM orders
+                WHERE customer_id = ${customerId}
+            `;
```
Comment (hualocson) [inline src/apis/controllers/orders.controller.js]: Query 1 time

File: src/apis/controllers/orders.controller.js [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+                FROM orders
+                WHERE customer_id = ${customerId} AND status = ${status}
+            `;
+      return successResponse(
+        { orders },
+        "Get orders by customerId and status successfully",
+        200
+      );
+    }
+  ),
```
Comment (hualocson) [inline src/apis/controllers/orders.controller.js]: Remove `getOrdersByCustomerIdAndStatus` endpoint

File: src/apis/controllers/orders.controller.js [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+      if (!order) {
+        return errorResponse(`Order with id ${id} not found`, 404);
+      }
+      // Check if status existed
+      const [statusEnum] = await sql`
+                SELECT enum_value FROM enums WHERE enum_value = ${status}
+            `;
+      if (!statusEnum) {
+        return errorResponse(`Status ${status} is invalid`, 400);
+      }
```
Comment (hualocson) [inline src/apis/controllers/orders.controller.js]: use validator in routes for check valid `status`

File: src/apis/controllers/orders.controller.js [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+                FROM orders
+                WHERE status = ${status}
+            `;
+      return successResponse(
+        { orders },
+        "Get orders by status successfully",
+        200
+      );
+    }
+  ),
```
Comment (hualocson) [inline src/apis/controllers/orders.controller.js]: Remove this endpoint

File: src/apis/routes/orders.routes.js [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    ordersController.getAllOrdersByCustomerId
+  );
+
+  // Get orders by status from req.params, status is a number
+  router.get(
+    "/orders/status/:status",
+    adminAuthorization(1),
+    param("status").isInt().withMessage("Status must be a number"),
+    ordersController.getOrdersByStatus
+  );
```
Comment (hualocson) [inline src/apis/routes/orders.routes.js]: Remove this

File: src/apis/routes/orders.routes.js [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+  // Get orders by both customerId and status in query not params
+  router.get(
+    "/orders/customer/:customerId/status/:status",
+    adminAuthorization(1),
+    param("customerId").isInt().withMessage("Id must be a number"),
+    param("status").isInt().withMessage("Status must be a number"),
+    ordersController.getOrdersByCustomerIdAndStatus
+  );
```
Comment (hualocson) [inline src/apis/routes/orders.routes.js]: Remove this

File: src/apis/routes/orders.routes.js [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+  // Update only status of order
+  router.patch(
+    "/orders/update-status/:id",
+    adminAuthorization(2),
+    param("id").isInt().withMessage("Id must be a number"),
+    body("status").isInt().withMessage("Status must be a number"),
+    ordersController.updateOrder
+  );
```
Comment (hualocson) [inline src/apis/routes/orders.routes.js]: change path to `/orders/:id/status`

**Review by hualocson — CHANGES_REQUESTED** [review]

File: src/apis/controllers/orders.controller.js (hunk 1 above)
Comment (binh0804) [reply src/apis/controllers/orders.controller.js]: Done

File: src/apis/routes/orders.routes.js (hunk 2 above)
Comment (binh0804) [reply src/apis/routes/orders.routes.js]: Done

File: src/apis/controllers/orders.controller.js (hunk 3 above)
Comment (binh0804) [reply src/apis/controllers/orders.controller.js]: Done

File: src/apis/controllers/orders.controller.js (hunk 4 above)
Comment (binh0804) [reply src/apis/controllers/orders.controller.js]: Done

File: src/apis/controllers/orders.controller.js (hunk 5 above)
Comment (binh0804) [reply src/apis/controllers/orders.controller.js]: Done

**Review by hualocson — APPROVED** [review]

**Pull request merged**